### ISSUE TITLE
please make a flatpak package for linux

### ISSUE THREAD

**Issue by CamJRP** [coelckers/Raze#403]
I know that this is asking too much, but could anyone please make a Flatpak package in the near future.
That would help a lot to the Linux adoption of this awesome source port.

**Comment by Cacodemon345** [comment]
I don't think Flatpak is the right solution here. AppImage is, since you can dynamically link with system GTK and thus don't run into issues with theme mismatch due to differing GTK versions.

**Comment by coelckers** [comment]
Please let's not start philosophical debates about which app distribution format is better. From past experience I can tell that these discussions never end well because too much personal bias gets into the discussion. One more such post and they'll all get deleted!

**Comment by CamJRP** [comment]
> Please let's not start philosophical debates about which app distribution format is better. From past experience I can tell that these discussions never end well because too much personal bias gets into the discussion. One more such post and they'll all get deleted!

I don't think this is a philosophical debate, i'm just proposing a way to facilitate linux adoption of the source port. It could be flatpak or appimage, i really don't care

**Comment by coelckers** [comment]
I was addressing Cacodemon who brought in that "opinion" nonsense.

**Comment by Cacodemon345** [comment]
TBH, I don't really care. It's ultimately going to be the developers' decision about whether to make use of Linux app packaging and if so, which format to use.

**Comment by Gaerzi** [comment]
As far as I know, there's no obstacle to supporting both Flatpak and AppImage, so supporting Flatpak now doesn't preclude also supporting AppImage later. It's not really a choice to make. (Cue Mia Agraviador.)

**Comment by Calinou** [comment]
For reference, GZDoom already has a Flatpak: https://flathub.org/apps/details/org.zdoom.GZDoom

It shouldn't be too difficult to do the same for Raze.

**Comment by Calinou** [comment]
Some progress: The Flatpak builds and runs, but it can't find game data, even if it install it to `~/.var/app/org.zdoom.Raze/.config/raze/duke/DUKE3D.GRP` (with `DUKE.RTS` besides it).

I've tesdted the GZDoom Flatpak which uses the same logic to sandbox and place files, and it works there.

**Comment by CamJRP** [comment]
> Some progress: The [Flatpak](https://github.com/Calinou/org.zdoom.Raze) builds and runs, but it can't find game data, even if I install it to `~/.var/app/org.zdoom.Raze/.config/raze/duke/DUKE3D.GRP` (with `DUKE.RTS` besides it). I've also tried command line arguments to force specific paths to no avail (even when `--filesystem=host` is added).
> 
> Note that the expected path is `~/.var/app/org.zdoom.Raze/.config` with `.config`, not `config`. (I'm not sure why this is the case.)
> 
> I've tested the GZDoom Flatpak which uses the same logic to sandbox and place files, and it works there.
> 
> cc @Eonfge

Thanks a lot. 
Maybe Flatseal could solve the problem.

**Comment by Unrud** [comment]
> Some progress: The [Flatpak](https://github.com/Calinou/org.zdoom.Raze) builds and runs, but it can't find game data

It can't find **raze.pk3**. The path `/app/share/games/raze` is not in **FileSearch.Directories**.

The paths that are added to the default configuration are hard-coded here:

https://github.com/coelckers/Raze/blob/b6eadbccb5c733c7710ca8651e8e934a276eab6a/source/core/gameconfigfile.cpp#L135-L142

> Note that the expected path is `~/.var/app/org.zdoom.Raze/.config` with `.config`, not `config`. (I'm not sure why this is the case.)

Because `XDG_CONFIG_HOME` is not respected. It uses `$HOME/.config`.

The wrapper script with `raze +fluid_patchset /app/share/sounds/sf2/raze.sf2 ...` is not required if you add `/app/share/sounds/sf2` to **SoundfontSearch.Directories**.

**Comment by mjr4077au** [comment]
> Some progress: The [Flatpak](https://github.com/Calinou/org.zdoom.Raze) builds and runs, but it can't find game data, even if I install it to `~/.var/app/org.zdoom.Raze/.config/raze/duke/DUKE3D.GRP` (with `DUKE.RTS` besides it). I've also tried command line arguments to force specific paths to no avail (even when `--filesystem=host` is added).
> 
> Note that the expected path is `~/.var/app/org.zdoom.Raze/.config` with `.config`, not `config`. (I'm not sure why this is the case.)
> 
> I've tested the GZDoom Flatpak which uses the same logic to sandbox and place files, and it works there.
> 
> cc @Eonfge

@Calinou if you didn't mind adding this again, I added some code to specify SYSTEMINSTALL (i.e. it's a packaged app) in 0033826544c8905883283e5193cb5cdaed9c80c8. Between that, overridding `CMAKE_INSTALL_PREFIX` and overriding `INSTALL_PK3_PATH`, you should be able fix your "couldn't find data" issue, hopefully.

**Comment by Calinou** [comment]
@mjr4077au Thanks!

I've pushed an update: https://github.com/Calinou/org.zdoom.Raze

However, Raze still loads game data from my `$HOME/.config/raze` (when it should be using the Flatpak-specific `$HOME/.var/app/org.zdoom.Raze/config/raze` instead). Also, the game crashes as soon as it tries to play music:

```
Resolution: 640 x 480
player 1 of 1 (1 nodes)
Could not find patch set raze.

Could not find patch set /usr/share/sounds/sf2/FluidR3_GS.sf2.

Could not find patch set /usr/share/sounds/sf2/FluidR3_GM.sf2.



*** Fatal Error ***
!!! Failed to exec debug process**
```

I did install the Raze soundfont during the Flatpak build process, but there's no `FluidR3_GS.sf2` and `FluidR3_GM.sf2` to install.

**Comment by mjr4077au** [comment]
You're going to have to patch the locations in gameconfigfile.cpp to remove what you don't want and add what you do want.

For ZMusic, you'll have to patch where you're storing the soundfont. A similar patch used in Arch Linux is here: https://aur.archlinux.org/cgit/aur.git/tree/0001-Use-correct-soundfont-path.patch?h=zmusic-git

**Comment by parherman** [comment]
@Calinou 

Maybe paths in source/common/platform/posix/unix/i_specialpaths.cpp needs to be patched as well?

Or maybe you could use configure --prefix $HOME? (found in the ChangeLog)

I made a test build with your repo + update commits, updated runtime and ```- --filesystem=xdg-config/raze:create``` and it runs fine here!

**Comment by smirnov** [comment]
I've updated the version and built myself a working flatpak version here: https://github.com/smirnov/org.zdoom.Raze

Would any of the devs object if I put it up on Flathub?

**Comment by smirnov** [comment]
Just as a small update, Raze is now available on Flathub, I hope noone minds, if any of the devs have anything against that or would like to be owner/collaborator - please let me know. I'm happy to keep maintaining for the foreseeable future, but do not intend to "hijack" the project representation.

**Comment by Eonfge** [comment]
I've never actually played Duke Nukem so perhaps this is the excuse I need. Either way, it's cool that somebody finally had the time and energy to see this though. Congratulations.

The hardest part is done, it now comes down to maintenance.

**Comment by bobberb** [comment]
@smirnov thank you for your effort in stepping up to maintain. I have already placed a new Issue in your honor :D !

**Comment by CamJRP** [comment]
Thanks a lot to everyone involved in supporting the flatpak. 
I just installed it in my debian machine, but is not able to find the game data. 
Anyone knows how to fix this problem?

**Comment by smirnov** [comment]
You can place the data in ~/.var/app/org.zdoom.Raze/.config/raze/
or use
flatpak override org.zdoom.Raze --filesystem=/OTHER/LOCATION/WITH/GRPS --user

**Comment by Calinou** [comment]
@coelckers This issue can be closed now, as issues specific to the Flatpak distribution should be reported on https://github.com/flathub/org.zdoom.Raze.
